hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/AttributeSearchService.kt

```diff
@@ -0,0 +1,150 @@
+
+import jakarta.validation.ValidationException
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.prisonersearch.common.model.Prisoner
+import uk.gov.justice.digital.hmpps.prisonersearch.search.resource.PrisonerSearchResource
+import uk.gov.justice.digital.hmpps.prisonersearch.search.services.dto.AttributeSearchRequest
+import uk.gov.justice.digital.hmpps.prisonersearch.search.services.dto.Matcher
+import uk.gov.justice.digital.hmpps.prisonersearch.search.services.dto.TextMatcher
+import java.time.LocalDate
+import java.time.LocalDateTime
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.memberProperties
+
+enum class AttributeType {
+  STRING,
+  INTEGER,
+  BOOLEAN,
+  DATE,
+  DATE_TIME,
+}
+
+@Component
+class AttributeSearchService {
+
+  private val attributeTypes: Map<String, AttributeType> = AttributeTypeResolver().getAttributeTypes(Prisoner::class)
+
+  fun search(request: AttributeSearchRequest) {
+    validate(request)
+    log.info("searchByAttributes called with request: $request")
+  }
+
+  fun validate(request: AttributeSearchRequest) {
+    if (request.matchers.isEmpty()) {
+      throw AttributeSearchException("At least one matcher must be provided")
+    }
+    getAllMatchers(request.matchers).forEach { validateMatcher(it) }
+    getAllTextMatchers(request.matchers).forEach { validateTextMatcher(it) }
+  }
+
+  private fun validateMatcher(matcher: Matcher) {
+    if (matcher.textMatchers.isNullOrEmpty() &&
+      matcher.booleanMatchers.isNullOrEmpty() &&
+      matcher.integerMatchers.isNullOrEmpty() &&
+      matcher.dateMatchers.isNullOrEmpty() &&
+      matcher.bodyPartMatchers.isNullOrEmpty() &&
+      matcher.children.isNullOrEmpty()
+    ) {
+      throw AttributeSearchException("Matchers must not be empty")
+    }
+  }
+
+  private fun getAllMatchers(matcher: List<Matcher>): List<Matcher> {
+    val allMatchers = mutableListOf<Matcher>()
+    matcher.forEach {
+      allMatchers.add(it)
+      it.children?.let { children -> allMatchers.addAll(getAllMatchers(children)) }
+    }
+    return allMatchers
+  }
+
+  private fun getAllTextMatchers(matcher: List<Matcher>): List<TextMatcher> {
+    val allTextMatchers = mutableListOf<TextMatcher>()
+    matcher.forEach {
+      it.textMatchers?.let { textMatchers -> allTextMatchers.addAll(textMatchers) }
+      it.children?.let { children -> allTextMatchers.addAll(getAllTextMatchers(children)) }
+    }
+    return allTextMatchers
+  }
+
+  private fun validateTextMatcher(textMatcher: TextMatcher) {
+    attributeTypes[textMatcher.attribute]
+      ?.also {
+        if (it != AttributeType.STRING) {
+          throw AttributeSearchException("Attribute ${textMatcher.attribute} is not a text attribute")
+        }
+      }
+      ?: throw AttributeSearchException("Unknown attribute: ${textMatcher.attribute}")
```

Comment:
We only support attributes that we found on the Prisoner record. Currently only implemented the text attribute validation but this check might move up as it will be needed for all types.